x/ccv/provider/keeper/distribution.go

```diff
@@ -57,32 +68,159 @@ func (k Keeper) GetAllConsumerRewardDenoms(ctx sdk.Context) (consumerRewardDenom
+		return
+	}
 
-	// 2. Iterate over the whitelist
-	for _, denom := range denoms {
-		// 3. For each denom, retrieve the balance from the consumer rewards pool
-		balance := k.bankKeeper.GetBalance(
+	// Iterate over all registered consumer chains
+	for _, consumer := range k.GetAllConsumerChains(ctx) {
+
+		// transfer the consumer rewards to the distribution module account
+		// note that the rewards transferred are only consumer whitelisted denoms
+		rewardsCollected, err := k.TransferConsumerRewardsToDistributionModule(ctx, consumer.ChainId)
+		if err != nil {
+			panic(err)
+		}
+
+		if rewardsCollected.IsZero() {
+			continue
+		}
+
+		rewardsCollectedDec := sdk.NewDecCoinsFromCoins(rewardsCollected...)
+
+		// temporary workaround to keep CanWithdrawInvariant happy
+		// general discussions here: https://github.com/cosmos/cosmos-sdk/issues/2906#issuecomment-441867634
+		feePool := k.distributionKeeper.GetFeePool(ctx)
+		if totalPreviousPower == 0 {
+			feePool.CommunityPool = feePool.CommunityPool.Add(rewardsCollectedDec...)
+			k.distributionKeeper.SetFeePool(ctx, feePool)
+			return
+		}
+
+		// Calculate the reward allocations
+		remaining := rewardsCollectedDec
+		communityTax := k.distributionKeeper.GetCommunityTax(ctx)
+		voteMultiplier := math.LegacyOneDec().Sub(communityTax)
+		feeMultiplier := rewardsCollectedDec.MulDecTruncate(voteMultiplier)
+
+		// allocate tokens to consumer validators
+		feeAllocated := k.AllocateTokensToConsumerValidators(
+			ctx,
+			consumer.ChainId,
+			totalPreviousPower,
+			bondedVotes,
+			feeMultiplier,
+		)
+		remaining = remaining.Sub(feeAllocated)
+
+		// allocate community funding
+		feePool.CommunityPool = feePool.CommunityPool.Add(remaining...)
+		k.distributionKeeper.SetFeePool(ctx, feePool)
+	}
+}
+
+// TODO: allocate tokens to validators that opted-in and for long enough e.g. 1000 blocks
+// once the opt-in logic is integrated QueueVSCPackets()
+//
+// AllocateTokensToConsumerValidators allocates the consumer rewards pool to validator
+// according to their voting power
+func (k Keeper) AllocateTokensToConsumerValidators(
+	ctx sdk.Context,
+	chainID string,
+	totalPreviousPower int64,
+	bondedVotes []abci.VoteInfo,
+	tokens sdk.DecCoins,
+) sdk.DecCoins {
+	totalReward := sdk.DecCoins{}
+	for _, vote := range bondedVotes {
+		// TODO: should check if validator IsOptIn or continue here
+		consAddr := sdk.ConsAddress(vote.Validator.Address)
+
+		// TODO: Consider micro-slashing for missing votes.
+		//
+		// Ref: https://github.com/cosmos/cosmos-sdk/issues/2525#issuecomment-430838701
+		powerFraction := math.LegacyNewDec(vote.Validator.Power).QuoTruncate(math.LegacyNewDec(totalPreviousPower))
+		tokensFraction := tokens.MulDecTruncate(powerFraction)
+
+		k.distributionKeeper.AllocateTokensToValidator(
 			ctx,
-			k.accountKeeper.GetModuleAccount(ctx, types.ConsumerRewardsPool).GetAddress(),
-			denom,
+			k.stakingKeeper.ValidatorByConsAddr(ctx, consAddr),
```

Comment:
Consider tweaking the validator's commission here if it is possible. This would involve storing a mapping of validator/consumer chain -> commission, so that validators could have custom commissions per consumer chain. This would be an excellent feature to add.